OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs

```diff
@@ -52,7 +52,7 @@ void ITick.Tick(Actor self)
 				return;
 
 			var currentTerrain = self.World.Map.GetTerrainInfo(self.Location).Type;
```

Comment:
Other layers can actually define their own terrain types, so a more comprehensive fix would be something like:
```csharp
var currentTerrain = self.Location.Layer == 0 ? self.World.Map.GetTerrainInfo(self.Location).Type :
		self.World.Map.Rules.TileSet[self.World.GetCustomMovementLayers()[self.Location.Layer].Index].Type;
```

Reply:
I tried your approach but it returns on
bridges `DirtyRoad` instead of `Road` and in
tunnels `Bridge`.

What do you mean by cache appropriate references? Storing the last cell (before entering tunnel/bridge) and working with that Custom Layer instead of the Actors current?

Reply:
Sorry, my "something like" wasn't quite right: the `.Index` should have been `.GetTerrainIndex(self.Location)`.